src/apphosting/yaml.ts

```diff
@@ -0,0 +1,133 @@
+import { basename, dirname } from "path";
+import { readFileFromDirectory, wrappedSafeLoad } from "../utils";
+import { Config, Env, store } from "./config";
+import * as yaml from "yaml";
+import * as jsYaml from "js-yaml";
+import { fileExistsSync } from "../fsutils";
+
+export type EnvironmentVariable = Omit<Env, "secret">;
+export type Secret = Omit<Env, "availability" | "value">;
+
+/**
+ * AppHostingYamlConfig is an object representing an Apphosting Yaml configuration
+ * present in the user's codebase (i.e 'apphosting.yaml', 'apphosting.staging.yaml', etc).
+ *
+ * This class is used to abstract away the parsing and handling logic of these
+ * yaml files.
+ */
+export class AppHostingYamlConfig {
+  filePath?: string;
+  _loadedAppHostingYaml: Config;
+  private _environmentVariables: Map<string, EnvironmentVariable>;
+  private _secrets: Map<string, Secret>;
+
+  constructor(filePath?: string) {
+    this.filePath = filePath;
+    this._loadedAppHostingYaml = {};
+    this._environmentVariables = new Map();
+    this._secrets = new Map();
+  }
+
+  async init() {
+    // If the file doesn't exist don't load it in, instead it will be used to write to later
+
+    if (this.filePath && fileExistsSync(this.filePath)) {
+      const file = await readFileFromDirectory(dirname(this.filePath), basename(this.filePath));
+      this._loadedAppHostingYaml = (await wrappedSafeLoad(file.source)) ?? {};
+
+      if (this._loadedAppHostingYaml.env) {
+        const parsedEnvs = this.parseEnv(this._loadedAppHostingYaml.env);
+        this._environmentVariables = parsedEnvs.environmentVariables;
+        this._secrets = parsedEnvs.secrets;
+      }
+    }
+  }
+
+  get environmentVariables(): EnvironmentVariable[] {
+    return this.mapToEnv(this._environmentVariables);
+  }
+
+  get secrets(): Secret[] {
+    return this.mapToEnv(this._secrets);
+  }
+
+  environmentVariablesAsRecord(): Record<string, string> {
+    const newObject: Record<string, string> = {};
+    for (const [key, env] of this._environmentVariables) {
+      newObject[key] = env.value!;
+    }
+    return newObject;
+  }
+
+  addEnvironmentVariable(env: EnvironmentVariable) {
+    this._environmentVariables.set(env.variable, env);
+  }
+
+  addSecret(secret: Secret) {
+    this._secrets.set(secret.variable, secret);
+  }
```

Comment:
thoughts on getting rid of these and letting clients access and mutate the fields directly?

config.secrets.set(key, val)

doesn't seem too bad to me

Reply:
I like that the setters hide implementation, and the getters do a transformation so that the values are in `Record<string, string>[]` format which is easier to use by clients, compared to a Map.